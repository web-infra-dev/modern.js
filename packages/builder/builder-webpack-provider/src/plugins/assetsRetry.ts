import { BuilderPlugin } from '../types';
import { isHtmlDisabled } from './html';

export function builderPluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();

        if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
          return;
        }

        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
