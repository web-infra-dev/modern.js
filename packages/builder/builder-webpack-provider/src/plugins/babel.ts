import {
  getBabelConfig,
  createBabelChain,
  type BabelOptions,
} from '@modern-js/babel-preset-app';
import {
  JS_REGEX,
  TS_REGEX,
  mergeRegex,
  createVirtualModule,
  getBrowserslistWithDefault,
  type BuilderContext,
} from '@modern-js/builder-shared';

import type { WebpackChain, BuilderPlugin, NormalizedConfig } from '../types';

const enableCoreJsEntry = (config: NormalizedConfig, isServer: boolean) =>
  config.output.polyfill === 'entry' && !isServer;

export const getUseBuiltIns = (config: NormalizedConfig) => {
  const { polyfill } = config.output;
  if (polyfill === 'ua' || polyfill === 'off') {
    return false;
  }
  return polyfill;
};

export function applyScriptCondition({
  rule,
  config,
  context,
  includes,
  excludes,
}: {
  rule: WebpackChain.Rule;
  config: NormalizedConfig;
  context: BuilderContext;
  includes: (string | RegExp)[];
  excludes: (string | RegExp)[];
}) {
  // compile all folders in app directory, exclude node_modules
  rule.include.add({
    and: [context.rootPath, { not: /node_modules/ }],
  });

  [...includes, ...(config.source.include || [])].forEach(condition => {
    rule.include.add(condition);
  });

  [...excludes, ...(config.source.exclude || [])].forEach(condition => {
    rule.exclude.add(condition);
  });
}

export const builderPluginBabel = (): BuilderPlugin => ({
  name: 'builder-plugin-babel',
  setup(api) {
    api.modifyWebpackChain(
      async (
        chain,
        { CHAIN_ID, getCompiledPath, target, isProd, isServer },
      ) => {
        const { applyOptionsChain, isUseSSRBundle } = await import(
          '@modern-js/utils'
        );

        const config = api.getNormalizedConfig();
        const browserslist = await getBrowserslistWithDefault(
          api.context.rootPath,
          config,
          target,
        );

        const getBabelOptions = (
          appDirectory: string,
          config: NormalizedConfig,
        ) => {
          // 1. Get styled-components options
          const styledComponentsOptions = applyOptionsChain(
            {
              // "pure" is used to improve dead code elimination in production.
              // we don't need to enable it in development because it will slow down the build process.
              pure: isProd,
              displayName: true,
              ssr: isUseSSRBundle(config),
              transpileTemplateLiterals: true,
            },
            config.tools.styledComponents,
          );

          // 2. Create babel util function about include/exclude
          const includes: Array<string | RegExp> = [];
          const excludes: Array<string | RegExp> = [];

          const babelUtils = {
            addIncludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                includes.push(...items);
              } else {
                includes.push(items);
              }
            },
            addExcludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                excludes.push(...items);
              } else {
                excludes.push(items);
              }
            },
          };

          // 3. Compute final babel config by @modern-js/babel-preset-app
          const babelOptions: BabelOptions = {
            babelrc: false,
            configFile: false,
            compact: isProd,
            ...getBabelConfig({
              target: isServer ? 'server' : 'client',
              appDirectory,
              useLegacyDecorators: !config.output.enableLatestDecorators,
              useBuiltIns: isServer ? false : getUseBuiltIns(config),
              chain: createBabelChain(),
              styledComponents: styledComponentsOptions,
              userBabelConfig: config.tools.babel,
              userBabelConfigUtils: babelUtils,
              overrideBrowserslist: browserslist,
            }),
          };

          if (config.output.charset === 'utf8') {
            babelOptions.generatorOpts = {
              jsescOption: { minimal: true },
            };
          }

          return {
            babelOptions,
            includes,
            excludes,
          };
        };

        const { rootPath } = api.context;
        const { babelOptions, includes, excludes } = getBabelOptions(
          rootPath,
          config,
        );
        const useTsLoader = Boolean(config.tools.tsLoader);
        const rule = chain.module.rule(CHAIN_ID.RULE.JS);

        applyScriptCondition({
          rule,
          config,
          context: api.context,
          includes,
          excludes,
        });

        rule
          .test(useTsLoader ? JS_REGEX : mergeRegex(JS_REGEX, TS_REGEX))
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelOptions);

        /**
         * If a script is imported with data URI, it can be compiled by babel too.
         * This is used by some frameworks to create virtual entry.
         * https://webpack.js.org/api/module-methods/#import
         * @example: import x from 'data:text/javascript,export default 1;';
         */
        if (config.source.compileJsDataURI) {
          chain.module
            .rule(CHAIN_ID.RULE.JS_DATA_URI)
            .mimetype({
              or: ['text/javascript', 'application/javascript'],
            })
            .use(CHAIN_ID.USE.BABEL)
            .loader(getCompiledPath('babel-loader'))
            .options(babelOptions);
        }

        addCoreJsEntry({ chain, config, isServer });
      },
    );
  },
});

/** Add core-js-entry to every entries. */
export function addCoreJsEntry({
  chain,
  config,
  isServer,
}: {
  chain: WebpackChain;
  config: NormalizedConfig;
  isServer: boolean;
}) {
  if (enableCoreJsEntry(config, isServer)) {
    const entryPoints = Object.keys(chain.entryPoints.entries() || {});
    const coreJsEntry = createVirtualModule('import "core-js";');

    for (const name of entryPoints) {
      chain.entry(name).prepend(coreJsEntry);
    }
  }
}
