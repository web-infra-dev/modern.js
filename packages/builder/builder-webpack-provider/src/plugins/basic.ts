import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of webpack
 */
export const builderPluginBasic = (): BuilderPlugin => ({
  name: 'builder-plugin-basic',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      chain.context(api.context.rootPath);

      chain.mode(isProd ? 'production' : 'development');

      chain.merge({
        infrastructureLogging: {
          // Using `error` level to avoid `webpack.cache.PackFileCacheStrategy` logs
          level: 'error',
        },
      });

      // if the chunk size exceeds 1MiB, we will throw a warning
      chain.performance.maxAssetSize(1024 * 1024);
      chain.performance.maxEntrypointSize(1024 * 1024);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });
    });
  },
});
