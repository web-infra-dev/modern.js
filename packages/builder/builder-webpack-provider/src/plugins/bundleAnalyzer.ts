import { BuilderPlugin } from '../types';

export function builderPluginBundleAnalyzer(): BuilderPlugin {
  return {
    name: 'builder-plugin-bundle-analyzer',
    setup(api) {
      api.modifyWebpackChain(
        async (chain, { CHAIN_ID, getCompiledPath, target }) => {
          const config = api.getNormalizedConfig();
          // There are two ways to open the bundle analyzer:
          // 1. Set environment variable `BUNDLE_ANALYZE`
          // 2. Set performance.bundleAnalyze config
          if (
            !process.env.BUNDLE_ANALYZE &&
            !config.performance.bundleAnalyze
          ) {
            return;
          }
          const { default: BundleAnalyzer } = await import(
            getCompiledPath('webpack-bundle-analyzer')
          );
          chain
            .plugin(CHAIN_ID.PLUGIN.BUNDLE_ANALYZER)
            .use(BundleAnalyzer.BundleAnalyzerPlugin, [
              {
                analyzerMode: 'static',
                openAnalyzer: false,
                reportFilename: `report-${target}.html`,
                ...(config.performance.bundleAnalyze || {}),
              },
            ]);
        },
      );
    },
  };
}
