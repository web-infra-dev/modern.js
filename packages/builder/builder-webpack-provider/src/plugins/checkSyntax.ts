import {
  BuilderContext,
  BuilderTarget,
  DEFAULT_BROWSERSLIST,
  getBrowserslistWithDefault,
} from '@modern-js/builder-shared';
import type { BuilderPlugin, NormalizedConfig } from '../types';

export function builderPluginCheckSyntax(): BuilderPlugin {
  return {
    name: 'builder-plugin-check-syntax',

    setup(api) {
      api.modifyWebpackChain(async (chain, { isProd, target }) => {
        const config = api.getNormalizedConfig();
        const { checkSyntax } = config.security;

        if (
          !isProd ||
          ['node', 'web-worker'].includes(target) ||
          !checkSyntax
        ) {
          return;
        }

        const targets = await getCheckTargets(
          api.context,
          config,
          target,
          checkSyntax,
        );
        const { CheckSyntaxPlugin } = await import(
          '../webpackPlugins/CheckSyntaxPlugin'
        );
        chain.plugin(CheckSyntaxPlugin.name).use(CheckSyntaxPlugin, [
          {
            targets,
            exclude:
              typeof checkSyntax === 'object' ? checkSyntax.exclude : undefined,
          },
        ]);
      });
    },
  };
}

async function getCheckTargets(
  builderContext: BuilderContext,
  builderConfig: NormalizedConfig,
  builderTarget: BuilderTarget,
  checkSyntax: { targets: string[] } | true,
) {
  if (checkSyntax === true) {
    const browserslist = await getBrowserslistWithDefault(
      builderContext.rootPath,
      builderConfig,
      builderTarget,
    );

    return browserslist || DEFAULT_BROWSERSLIST[builderTarget];
  }
  return checkSyntax.targets;
}
