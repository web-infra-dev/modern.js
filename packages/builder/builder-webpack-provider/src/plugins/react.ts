import type { BuilderPlugin } from '../types';

export const PluginReact = (): BuilderPlugin => ({
  name: 'builder-plugin-react',

  setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID, isProd }) => {
      const config = api.getBuilderConfig();

      if (isProd || config.dev?.hmr === false) {
        return;
      }

      const { default: ReactFastRefreshPlugin } = await import(
        '@pmmmwh/react-refresh-webpack-plugin'
      );
      const useTsLoader = Boolean(config.tools?.tsLoader);
      const rule = useTsLoader
        ? chain.module.rule(CHAIN_ID.RULE.TS)
        : chain.module.rule(CHAIN_ID.RULE.JS);

      rule.use(CHAIN_ID.USE.BABEL).tap(options => ({
        ...options,
        plugins: [
          ...(options.plugins || []),
          [require.resolve('react-refresh/babel'), { skipEnvCheck: true }],
        ],
      }));

      chain
        .plugin(CHAIN_ID.PLUGIN.REACT_FAST_REFRESH)
        .use(ReactFastRefreshPlugin, [
          {
            overlay: false,
            exclude: [/node_modules/],
          },
        ]);
    });
  },
});
