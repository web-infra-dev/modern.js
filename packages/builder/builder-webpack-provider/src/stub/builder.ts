import type * as playwright from '@modern-js/e2e/playwright';
import { getTemplatePath } from '@modern-js/utils';
import _ from '@modern-js/utils/lodash';
import {
  createPluginStore,
  createPublicContext,
  applyDefaultBuilderOptions,
  type PluginStore,
  type BuildOptions,
  type CreateBuilderOptions,
} from '@modern-js/builder-shared';
import assert from 'assert';
import { PathLike } from 'fs';
import {
  applyBasicPlugins,
  applyDefaultPlugins,
  applyMinimalPlugins,
} from '../shared/plugin';
import { URL } from 'url';
import { Hooks } from '../core/initHooks';
import type { BuilderConfig, BuilderPlugin, Context } from '../types';
import { STUB_BUILDER_PLUGIN_BUILTIN } from './constants';
import { createStubContext } from './context';
import { matchLoader, globContentJSON, filenameToGlobExpr } from './utils';

export interface OptionsPluginsItem {
  builtin?: boolean | 'default' | 'minimal' | 'basic';
  additional?: BuilderPlugin[];
}

export interface StubBuilderOptions extends CreateBuilderOptions {
  context?: Context;
  /**
   * Setup builtin plugins and add custom plugins.
   * Automatically add builtin plugins by `process.env.STUB_BUILDER_PLUGIN_BUILTIN`.
   */
  plugins?: OptionsPluginsItem | OptionsPluginsItem[keyof OptionsPluginsItem];
  /**
   * Whether to run webpack build. By default it will be `false` and skip webpack building.
   * Set a string value to specify the `output.distPath` config.
   */
  webpack?: boolean | string;
  buildOptions?: BuildOptions;
  builderConfig?: BuilderConfig;
}

export type HookApi = {
  [key in keyof Hooks]: Parameters<Parameters<Hooks[key]['tap']>[0]>;
};

export interface ServeDestOptions {
  hangOn?: playwright.TestType<any, any> | boolean;
}

export function normalizeStubPluginOptions(
  options?: StubBuilderOptions['plugins'],
) {
  const opt: Required<OptionsPluginsItem> = {
    builtin: STUB_BUILDER_PLUGIN_BUILTIN ?? false,
    additional: [],
  };
  // normalize options
  if (Array.isArray(options)) {
    opt.additional = options;
  } else if (typeof options === 'object') {
    Object.assign(opt, options);
  } else if (typeof options === 'string' || typeof options === 'boolean') {
    opt.builtin = options;
  }
  return opt;
}

export async function applyPluginOptions(
  pluginStore: PluginStore,
  options?: StubBuilderOptions['plugins'],
) {
  const opt = normalizeStubPluginOptions(options);
  // apply plugins
  if (opt.builtin === true || opt.builtin === 'minimal') {
    pluginStore.addPlugins(await applyMinimalPlugins());
  } else if (opt.builtin === 'basic') {
    pluginStore.addPlugins(await applyBasicPlugins());
  } else if (opt.builtin === 'default') {
    pluginStore.addPlugins(await applyDefaultPlugins());
  }
  pluginStore.addPlugins(opt.additional);
}

/**
 * Create stub builder for testing.
 * Some behaviors will be different to common `createBuilder`.
 */
export async function createStubBuilder(options?: StubBuilderOptions) {
  // init primary builder.
  const builderOptions = applyDefaultBuilderOptions(
    options,
  ) as Required<StubBuilderOptions>;
  // apply webpack option.
  if (options?.webpack) {
    const distPath =
      typeof options.webpack === 'string'
        ? options.webpack
        : getTemplatePath('modern-js/stub-builder/dist');
    _.set(builderOptions.builderConfig, 'output.distPath', distPath);
  }
  // init context.
  const context = createStubContext(builderOptions);
  // merge user context.
  options?.context && _.merge(context, options.context);

  const publicContext = createPublicContext(context);
  const pluginStore = createPluginStore();

  // add builtin and custom plugins by `options.plugins`.
  await applyPluginOptions(pluginStore, options?.plugins);

  // tap on each hook and cache the args.
  const resolvedHooks: Record<string, any> = {};
  _.each(context.hooks, ({ tap }, name) => {
    tap((...args) => {
      resolvedHooks[name] = args;
    });
  });

  /**
   * Run build and caching args of all hooks.
   * It will run actual build only first time,
   * and always return cached result until {@link reset}.
   */
  const build = _.memoize(async () => {
    const { build: buildImpl, webpackBuild } = await import('../core/build');
    const executeBuild = options?.webpack ? webpackBuild : undefined;
    await buildImpl(
      { context, pluginStore, builderOptions },
      options?.buildOptions,
      executeBuild,
    );
    return { resolvedHooks: { ...resolvedHooks } };
  });

  /**
   * Unwrap args of hook.
   *
   * The `unwrap` means it will try to run {@link build} and run some asserts before return.
   */
  const unwrapHook = async <T extends keyof HookApi>(
    hook: T,
  ): Promise<HookApi[T]> => (await build()).resolvedHooks[hook];

  /** Unwrap webpack configs. */
  const unwrapWebpackConfigs = async () => {
    const [{ bundlerConfigs }] = await unwrapHook('onBeforeBuildHook');
    return bundlerConfigs;
  };

  /** Unwrap webpack config, it will ensure there's only one config object. */
  const unwrapWebpackConfig = async () => {
    const webpackConfigs = await unwrapWebpackConfigs();
    assert(webpackConfigs.length === 1);
    return webpackConfigs[0];
  };

  /** Unwrap webpack compiler instance. */
  const unwrapWebpackCompiler = async () => {
    const [{ compiler }] = await unwrapHook('onAfterCreateCompilerHooks');
    return compiler;
  };

  /** Serialize content of output files into JSON object. */
  const unwrapOutputJSON = async (
    paths: PathLike | PathLike[] = context.distPath,
    isRelative = false,
    maxSize = 4096,
  ) => {
    if (Array.isArray(paths) && isRelative) {
      throw new Error('`isRelative` is not supported for multiple paths.');
    }
    await build();
    const _paths = _(paths)
      .castArray()
      .map(filenameToGlobExpr)
      .map(String)
      .value();
    return globContentJSON(_paths, { absolute: !isRelative, maxSize });
  };

  /** Read output file content. */
  const readOutputFile = async (
    filename: string,
  ): Promise<string | undefined> => {
    const compiler = await unwrapWebpackCompiler();
    return new Promise((resolve, reject) => {
      compiler.outputFileSystem.readFile(filename, (err, data) =>
        err ? reject(err) : resolve(data?.toString('utf-8')),
      );
    });
  };

  /** Unwrap output file content, will ensure it is not empty. */
  const unwrapOutputFile = async (filename: string) => {
    await build();
    const content = await readOutputFile(filename);
    assert(content);
    return content;
  };

  /**
   * Run build and serve on `distPath`.
   * If `hangOn` is true, it will block running and let you debug the server.
   * Set `hangOn` to a {@link playwright.TestType} to automatically unset timeout of test.
   *
   * @example
   * test('should work', async () => {
   *   const builder = await createStubBuilder();
   *   const { baseurl } = builder.buildAndServe({ hangOn: test });
   *   page.goto(`${baseurl}/index.html`);
   * });
   */
  const buildAndServe = async (options?: ServeDestOptions) => {
    const [{ runStaticServer }] = await Promise.all([
      import('@modern-js/e2e'),
      build(),
    ]);
    const { port } = await runStaticServer(context.distPath);
    if (options?.hangOn) {
      // eslint-disable-next-line no-console
      console.log(
        `Successfully build, and hang on running server: http://localhost:${port}`,
      );
      typeof options.hangOn !== 'boolean' && options.hangOn.setTimeout(0);
      await new Promise(() => null);
    }
    // TODO: get real `htmlRoot` & `homeUrl` from compiler.
    const baseUrl = new URL(`http://localhost:${port}`);
    const htmlRoot = new URL('html/', baseUrl);
    const homeUrl = new URL('index/index.html', htmlRoot);
    return {
      baseUrl,
      htmlRoot,
      homeUrl,
      port,
    };
  };

  /** Match webpack plugin by constructor name. */
  const matchWebpackPlugin = async (pluginName: string) => {
    const config = await unwrapWebpackConfig();
    const result = config.plugins?.filter(
      item => item.constructor.name === pluginName,
    );
    if (Array.isArray(result)) {
      assert(result.length <= 1);
      return result[0] || null;
    } else {
      return result || null;
    }
  };

  /** Check if a file handled by specific loader. */
  const matchWebpackLoader = async (loader: string, testFile: string) =>
    matchLoader({ config: await unwrapWebpackConfig(), loader, testFile });

  /** Reset builder. @see {@link build} */
  const reset = () => {
    build.cache.clear!();
  };

  return {
    ...pluginStore,
    build,
    hooks: context.hooks,
    context,
    publicContext,
    unwrapHook,
    unwrapWebpackConfigs,
    unwrapWebpackConfig,
    unwrapWebpackCompiler,
    unwrapOutputJSON,
    unwrapOutputFile,
    readOutputFile,
    buildAndServe,
    matchWebpackPlugin,
    matchWebpackLoader,
    reset,
  };
}
