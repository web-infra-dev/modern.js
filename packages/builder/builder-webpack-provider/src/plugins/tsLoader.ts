import { TS_REGEX } from '@modern-js/builder-shared';
import { BuilderPlugin } from '../types';
import { applyScriptCondition, getUseBuiltIns } from './babel';

// Declare `ScriptTarget` and `ModuleKind` manualy to avoid high cost of typescript import
enum ScriptTarget {
  ESNext = 99,
}

enum ModuleKind {
  ESNext = 99,
}

export const PluginTsLoader = (): BuilderPlugin => {
  return {
    name: 'builder-plugin-ts-loader',
    setup(api) {
      const config = api.getBuilderConfig();
      if (!config.tools?.tsLoader) {
        return;
      }

      const { framework, rootPath } = api.context;
      const babelLoaderOptions = {
        presets: [
          [
            require.resolve('@modern-js/babel-preset-app'),
            {
              metaName: framework,
              appDirectory: rootPath,
              target: 'client',
              useTsLoader: true,
              useBuiltIns: getUseBuiltIns(config),
              userBabelConfig: config.tools.babel,
            },
          ],
        ],
      };

      const includes: Array<string | RegExp> = [];
      const excludes: Array<string | RegExp> = [];

      const tsLoaderUtils = {
        addIncludes(items: string | RegExp | (string | RegExp)[]) {
          if (Array.isArray(items)) {
            includes.push(...items);
          } else {
            includes.push(items);
          }
        },
        addExcludes(items: string | RegExp | (string | RegExp)[]) {
          if (Array.isArray(items)) {
            excludes.push(...items);
          } else {
            excludes.push(items);
          }
        },
      };
      api.modifyWebpackChain(async (chain, { getCompiledPath }) => {
        const { CHAIN_ID, applyOptionsChain } = await import(
          '@modern-js/utils'
        );
        const tsLoaderOptions = applyOptionsChain(
          {
            compilerOptions: {
              target: ScriptTarget.ESNext,
              module: ModuleKind.ESNext,
            },
            transpileOnly: true,
            allowTsInNodeModules: true,
          },
          config.tools?.tsLoader || {},
          tsLoaderUtils,
        );
        const rule = chain.module.rule(CHAIN_ID.RULE.TS);

        applyScriptCondition(rule, config, api.context, includes, excludes);

        rule
          .test(TS_REGEX)
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelLoaderOptions)
          .end()
          .use(CHAIN_ID.USE.TS)
          .loader(require.resolve('ts-loader'))
          .options(tsLoaderOptions);
      });
    },
  };
};
