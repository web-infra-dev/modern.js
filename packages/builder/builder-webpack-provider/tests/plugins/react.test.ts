import { expect, describe, it } from 'vitest';
import { builderPluginReact } from '@/plugins/react';
import { builderPluginBabel } from '@/plugins/babel';
import { builderPluginTsLoader } from '@/plugins/tsLoader';
import { createStubBuilder } from '@/stub/builder';

describe('plugins/react', () => {
  it('should work with babel-loader', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBabel(), builderPluginReact()],
      builderConfig: {
        tools: {
          tsChecker: false,
        },
      },
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should work with ts-loader', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginTsLoader(), builderPluginReact()],
      builderConfig: {
        tools: {
          tsLoader: {},
        },
      },
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should not apply react refresh when dev.hmr is false', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginReact()],
      builderConfig: {
        dev: {
          hmr: false,
        },
      },
    });

    expect(await builder.matchWebpackPlugin('ReactRefreshPlugin')).toBeFalsy();
  });

  it('should not apply react refresh when target is node', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginReact()],
      target: 'node',
    });

    expect(await builder.matchWebpackPlugin('ReactRefreshPlugin')).toBeFalsy();
  });

  it('should not apply react refresh when target is web-worker', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginReact()],
      target: 'web-worker',
    });

    expect(await builder.matchWebpackPlugin('ReactRefreshPlugin')).toBeFalsy();
  });
});
