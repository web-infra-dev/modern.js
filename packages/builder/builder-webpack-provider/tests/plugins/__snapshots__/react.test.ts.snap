// Vitest Snapshot v1

exports[`plugins/react > should work with babel-loader 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /node_modules/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<WORKSPACE>/packages/cli/babel-preset-app/src/built-in/babel-plugin-lock-corejs-version",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-dynamic-import-node/index.js",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-import/index.js",
                  {
                    "libraryDirectory": "es",
                    "libraryName": "antd",
                    "style": true,
                  },
                  "import-antd",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-lodash/index.js",
                  {},
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-decorators/index.js",
                  {
                    "legacy": true,
                  },
                ],
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "helpers": false,
                    "regenerator": true,
                    "useESModules": true,
                    "version": "7.18.6",
                  },
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-export-default-from/index.js",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-pipeline-operator/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-partial-application/index.js",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-styled-components/index.js",
                  {
                    "displayName": true,
                    "pure": false,
                    "ssr": false,
                    "transpileTemplateLiterals": true,
                  },
                  "styled-components",
                ],
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/react-refresh/babel.js",
                  {
                    "skipEnvCheck": true,
                  },
                ],
              ],
              "presets": [
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": false,
                    "corejs": {
                      "proposals": true,
                      "version": "3.26",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": false,
                    "shippedProposals": false,
                    "targets": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-react/lib/index.js",
                  {
                    "development": true,
                    "runtime": "classic",
                    "useBuiltIns": true,
                    "useSpread": false,
                  },
                ],
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<WORKSPACE>/packages/cli/babel-preset-app/src/built-in/babel-plugin-lock-corejs-version",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-dynamic-import-node/index.js",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-import/index.js",
                  {
                    "libraryDirectory": "es",
                    "libraryName": "antd",
                    "style": true,
                  },
                  "import-antd",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-lodash/index.js",
                  {},
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-decorators/index.js",
                  {
                    "legacy": true,
                  },
                ],
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "helpers": false,
                    "regenerator": true,
                    "useESModules": true,
                    "version": "7.18.6",
                  },
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-export-default-from/index.js",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-pipeline-operator/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-partial-application/index.js",
                ],
                [
                  "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-styled-components/index.js",
                  {
                    "displayName": true,
                    "pure": false,
                    "ssr": false,
                    "transpileTemplateLiterals": true,
                  },
                  "styled-components",
                ],
              ],
              "presets": [
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": false,
                    "corejs": {
                      "proposals": true,
                      "version": "3.26",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": false,
                    "shippedProposals": false,
                    "targets": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-react/lib/index.js",
                  {
                    "development": true,
                    "runtime": "classic",
                    "useBuiltIns": true,
                    "useSpread": false,
                  },
                ],
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "plugins": [
    ReactRefreshPlugin {
      "options": {
        "exclude": [
          /node_modules/,
        ],
        "include": /\\\\\\.\\(\\[cm\\]js\\|\\[jt\\]sx\\?\\|flow\\)\\$/i,
        "overlay": false,
      },
    },
  ],
}
`;

exports[`plugins/react > should work with ts-loader 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /node_modules/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/compiled/babel-loader",
            "options": {
              "plugins": [
                [
                  "<WORKSPACE>/node_modules/<PNPM_INNER>/react-refresh/babel.js",
                  {
                    "skipEnvCheck": true,
                  },
                ],
              ],
              "presets": [
                [
                  "<WORKSPACE>/packages/cli/babel-preset-app/dist/js/node/index.js",
                  {
                    "appDirectory": "<ROOT>",
                    "target": "client",
                    "useBuiltIns": "entry",
                    "useTsLoader": true,
                    "userBabelConfig": undefined,
                  },
                ],
              ],
            },
          },
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/ts-loader/index.js",
            "options": {
              "allowTsInNodeModules": true,
              "compilerOptions": {
                "module": 99,
                "target": 99,
              },
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": [
    ReactRefreshPlugin {
      "options": {
        "exclude": [
          /node_modules/,
        ],
        "include": /\\\\\\.\\(\\[cm\\]js\\|\\[jt\\]sx\\?\\|flow\\)\\$/i,
        "overlay": false,
      },
    },
  ],
}
`;
